libr/debug/trace.c

```diff
@@ -42,32 +42,32 @@ R_API int r_debug_trace_tag (RDebug *dbg, int tag) {
 	return (dbg->trace->tag = (tag>0)? tag: UT32_MAX);
 }
 
-R_API int r_debug_trace_pc (RDebug *dbg) {
+/*
+ * something happened at the given pc that we need to trace
+ */
+R_API int r_debug_trace_pc (RDebug *dbg, ut64 pc) {
 	ut8 buf[32];
-	RRegItem *ri;
 	RAnalOp op;
 	static ut64 oldpc = 0LL; // Must trace the previously traced instruction
-	r_debug_reg_sync (dbg, R_REG_TYPE_GPR, false);
-	if ((ri = r_reg_get (dbg->reg, dbg->reg->name[R_REG_NAME_PC], -1))) {
-		ut64 addr = r_reg_get_value (dbg->reg, ri);
-		if (!addr) {
-			return false;
-		}
-		if (dbg->iob.read_at (dbg->iob.io, addr, buf, sizeof (buf))>0) {
-			if (r_anal_op (dbg->anal, &op, addr, buf, sizeof (buf))>0) {
-				if (oldpc!=0LL) {
-					if (dbg->anal->esil) {
-						if (dbg->anal->trace) {
-							r_anal_esil_trace (dbg->anal->esil, &op);
-						}
+
+	if (dbg->iob.read_at (dbg->iob.io, pc, buf, sizeof (buf)) > 0) {
+		if (r_anal_op (dbg->anal, &op, pc, buf, sizeof (buf)) > 0) {
+			if (oldpc != 0LL) {
+				if (dbg->anal->esil) {
+					if (dbg->anal->trace) {
+						r_anal_esil_trace (dbg->anal->esil, &op);
 					}
-					r_debug_trace_add (dbg, oldpc, op.size);
 				}
-				oldpc = addr;
-				return true;
-			} else eprintf ("trace_pc: cannot get opcode size at 0x%"PFMT64x"\n", addr);
-		} //else eprintf ("trace_pc: cannot read memory at 0x%"PFMT64x"\n", addr);
-	} else eprintf ("trace_pc: cannot get program counter\n");
+				r_debug_trace_add (dbg, oldpc, op.size);
+			}
+			oldpc = pc;
+			return true;
+		}
```

Comment:
agree
> On 21 Jun 2016, at 00:58, Joshua J. Drake <notifications@github.com> wrote:
> 
> In libr/debug/trace.c <https://github.com/radare/radare2/pull/5185#discussion_r67782089>:
> 
> >  				}
> > -				oldpc = addr;
> > -				return true;
> > -			} else eprintf ("trace_pc: cannot get opcode size at 0x%"PFMT64x"\n", addr);
> > -		} //else eprintf ("trace_pc: cannot read memory at 0x%"PFMT64x"\n", addr);
> > -	} else eprintf ("trace_pc: cannot get program counter\n");
> > +				r_debug_trace_add (dbg, oldpc, op.size);
> > +			}
> > +			oldpc = pc;
> > +			return true;
> > +		}
> I'd prefer to invert some conditions and make this more readable that way. Nesting like this sucks for readability.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/radare/radare2/pull/5185/files/e9879b72dc00a8bffb17ddb925ed028cf4920275#r67782089>, or mute the thread <https://github.com/notifications/unsubscribe/AA3-lo3aKuzXx2aUe0D6_Q4JGIK3bP1qks5qNxsVgaJpZM4I6ILO>.
>